File: kernel/c/src/convert.c
//	convert.c
//	this file originally belonged to baseOS project
//		an OS template on which to build


#pragma once

#include "./lib/convert.h"

void to_hex(char *str, void *ptr) {
	str[PTR_CONVERT_LEN - 1] = '\0';
	str[0] = '0';
	str[1] = 'x';

	u8 *p = (u8 *) &ptr;
	for (size_t i = PTR_CONVERT_LEN - 2; i > 1; p++) {
		str[i--] = hex_literals[(*p & 0xf)];
		str[i--] = hex_literals[((*p >> 4) & 0xf)];
	}
}

void to_hexs(char* str, size_t h) {
	if (h == 0) {
		str[0] = '0';
		str[1] = '\0';
		return;
	}
	for (ssize_t i = U64_HEX_CONVERT_LEN-2; i >= 0; i--) {
		str[i] = hex_literals[(h >> (i*4)) & 0xf];
	}
	for (ssize_t i = U64_HEX_CONVERT_LEN - 2; i >= 0; i--) {
		if (str[i] != '0') {
			str[i+1] = '\0';
			break;
		}
	}
}



void to_string(char *str, size_t val) {
	if (val == 0) {
		str[0] = '0';
		str[1] = '\0';
	} else {
		size_t i = 0;
		while ((val > 0) && (i + 1 < U64_CONVERT_LEN)) {
			str[i++] = '0' + (val % 10);
			val /= 10;
		}
		str[i] = '\0';
		strrev(str, i);
	}
}

void to_stringi(char *str, ssize_t val) {
	if (val < 0){
		val = -val;
		str[0] = '-';
		str++;
	}
	to_string(str, val);
}


size_t to_int(const char* str) {
	ssize_t len = strlen(str);
	ssize_t i = 0;
	size_t result = 0, mul = 1;
	for (; (i < len) && ((str[i] >= '0') && (str[i] <= '9')); i++);
	--i;

	for (; (i >= 0) && ((str[i] >= '0') && (str[i] <= '9')); i--) {
		result += (str[i] - '0') * mul;
		mul *= 10;
	}

	return result;
}

ssize_t to_inti(const char* str) {
	ssize_t len = strlen(str);
	ssize_t i = 0;
	ssize_t result = 0;
	size_t mul = 1;
	if (str[0] == '-') {
		i++;
	}
	for (; (i < len) && ((str[i] >= '0') && (str[i] <= '9')); i++);
	--i;

	for (; (i >= 0) && ((str[i] >= '0') && (str[i] <= '9')); i--) {
		result += (str[i] - '0') * mul;
		mul *= 10;
	}
	if (str[i] == '-') {
		result = -result;
	}

	return result;
}

File: kernel/rust/src/convert.rs
//	convert.rs
//	this file originally belonged to baseOS project
//		an OS template on which to build

use crate::renderer::{Render, RENDERER};


pub trait Convert: Sized {
    type Type;
    fn into(&self) -> Self::Type;
    fn from(_: Self::Type) -> Self;
    fn from_ref(_: &Self::Type) -> Self;
}

pub trait ConvertMut<T>: Sized {
    type Type;
    type Error;
    fn mut_into(&self, data: &mut Self::Type) -> Result<(), Self::Error>;
    fn mut_from(&mut self, data: &Self::Type) -> Result<(), Self::Error>;
}

impl Render for u8 {
    #[inline(always)]
    fn render(&self) {
        unsafe {RENDERER.acquire_mut().acquire_mut().render(*self)};
    }
}


macro_rules! implement_render_for_unsigned {
    ($type:ty, $ln:expr) => {
        impl Render for $type {
            fn render(&self) {
                let rend = unsafe {crate::renderer::RENDERER.acquire_mut().acquire_mut()};
                if *self == 0 {
                    rend.render(b'0');
                }
                let mut num = *self;
                let mut index: isize = 0;
                let mut arr = [0u8; $ln];

                loop {
                    arr[index as usize] = (num % 10) as u8 + b'0';
                    num /= 10;
                    if num == 0 { break; }
                    index += 1;
                }

                while index >= 0 {
                    rend.render(arr[index as usize]);
                    index -= 1;
                }
            }
        }
    };
}

macro_rules! implement_render_for_signed {
    ($type:ty, $ln:expr) => {
        impl Render for $type {
            fn render(&self) {
                let rend = unsafe {RENDERER.acquire_mut().acquire_mut()};
                if *self == 0 {
                    rend.render(b'0');
                }
                let mut num = *self;
                if num < 0 {
                    num = -num;
                    rend.render(b'-');
                }
                let mut index: isize = 0;
                let mut arr = [0u8; $ln];

                loop {
                    arr[index as usize] = (num % 10) as u8 + b'0';
                    num /= 10;
                    if num == 0 { break; }
                    index += 1;
                }

                while index >= 0 {
                    rend.render(arr[index as usize]);
                    index -= 1;
                }
            }
        }
    };
}


implement_render_for_signed!(i8, 4);

implement_render_for_unsigned!(u16, 5);
implement_render_for_signed!(i16, 6);

implement_render_for_unsigned!(u32, 10);
implement_render_for_signed!(i32, 11);

implement_render_for_unsigned!(u64, 19);
implement_render_for_signed!(i64, 20);
implement_render_for_unsigned!(usize, 19);
implement_render_for_signed!(isize, 20);